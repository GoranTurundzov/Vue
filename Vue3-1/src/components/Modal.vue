<template>
    <div @click.self="closeModal" class="modal out">
        <div class="modal-dialog modal-dialog-centered">
            <div class="modal-content">
                <div class="block-header bg-primary-darker">
                    <ul class="block-options">
                        <li>
                            <button type="button" @click="closeModal" data-testid="x-button">
                                <i class="fa fa-times"></i>
                            </button>
                        </li>
                    </ul>
                    <slot name="header"></slot>
                </div>
                <div class="block-content">
                    <slot></slot>
                </div>
                <div class="modal-footer" v-if="footerSlotPassed">
                    <slot name="footer"></slot>
                </div>
            </div>
        </div>
    </div>
</template>

<script>
export default {
    name: 'Modal',
    emits: ['close-modal'],
    computed: {
        footerSlotPassed() {
            return this.$slots['footer'];
        }
    },
    methods: {
        closeModal() {
            this.$emit('close-modal');
        }
    }
};
</script>

<style scoped>
.out {
    background-color: rgba(7, 7, 7, 0.7);
    display: block;
    z-index: 2;
}
.modal-dialog {
    max-height: 92vh;
    overflow: scroll;
    -ms-overflow-style: none;
    scrollbar-width: none;
}
.modal-dialog::-webkit-scrollbar {
    display: none;
}
</style>
