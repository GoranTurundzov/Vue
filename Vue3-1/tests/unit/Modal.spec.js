import { mount } from '@vue/test-utils';
import Modal from '@/components/Modal';

let spyCloseModalFn = jest.spyOn(Modal.methods, 'closeModal');

describe('SubmissionModal', () => {
    let wrapper;

    beforeEach(() => {
        wrapper = mount(Modal);
    });

    it('Matches snapshot', () => expect(wrapper.element).toMatchSnapshot());

    test('Emitts when clicked outside', async () => {
        await wrapper.find('.out').trigger('click');
        expect(wrapper.emitted('close-modal')).toBeTruthy();
        expect(wrapper.emitted('close-modal').length).toBe(1);
    });

    test('Emmits when clicked on X button', async () => {
        await wrapper.find('[data-testid="x-button"]').trigger('click');
        expect(spyCloseModalFn).toBeCalled();
        expect(wrapper.emitted('close-modal')).toBeTruthy();
        expect(wrapper.emitted('close-modal').length).toBe(1);
    });

    test('prints slot and title prop', () => {
        let slotWrapper = mount(Modal, {
            slots: {
                default: '<div class="test-content">this is in the slot</div>',
                header: '<h3>Test Title</h3>',
                footer: '<p>All rights reserved</p>'
            }
        });
        expect(slotWrapper.find('.block-content').html()).toContain(
            '<div class="test-content">this is in the slot</div>'
        );
        expect(slotWrapper.find('.block-header').html()).toContain('<h3>Test Title</h3>');
        expect(slotWrapper.find('.modal-footer').exists()).toBe(true);
        expect(slotWrapper.find('.modal-footer').html()).toContain('<p>All rights reserved</p>');
    });
});
