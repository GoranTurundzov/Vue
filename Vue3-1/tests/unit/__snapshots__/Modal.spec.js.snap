// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SubmissionModal Matches snapshot 1`] = `
<div
  class="modal out"
>
  <div
    class="modal-dialog modal-dialog-centered"
  >
    <div
      class="modal-content"
    >
      <div
        class="block-header bg-primary-darker"
      >
        <ul
          class="block-options"
        >
          <li>
            <button
              data-testid="x-button"
              type="button"
            >
              <i
                class="fa fa-times"
              />
            </button>
          </li>
        </ul>
        
        
      </div>
      <div
        class="block-content"
      >
        
        
      </div>
      <!--v-if-->
    </div>
  </div>
</div>
`;
